### ISSUE TITLE
Unit tests fail with GCC 14

### ISSUE THREAD

**Issue by AlexanderRichert-NOAA** [NOAA-EMC/NCEPLIBS-bufr#627]
Flagging this for now (based on my attempts at updating CI tests), I'll look into this more soon.

**Comment by jbathegit** [comment]
Hi @AlexanderRichert-NOAA - just checking in to see if you've had any time to look further into this and if there are any updates on this issue?

**Comment by AlexanderRichert-NOAA** [comment]
I've looked at it just a little further-- here's the output of the tests that fail on my local machine (ubuntu 24) with gcc 14 (nothing fails if I use gcc 12): [bufr-gcc14-failures.txt](https://github.com/user-attachments/files/19303441/bufr-gcc14-failures.txt)
I started to dig in to the outtest1 failures, and it seems to have something to do with the output file writing-- I can't find any differences in the underlying data. In outtest1.F90, if I write the output file and stop the code after the `gettagpr()` call, I can see that the files are different.
@jbathegit before I continue down this rabbit hole, is there anything about the set of failing tests that stands out in terms of what they have in common? The failing tests are outtest{1,11,12,13}_{4,d,8}.

**Comment by jbathegit** [comment]
Thanks, and unfortunately they haven't installed gcc14 on WCOSS2 yet (latest available version is gcc13.2) so I can't try this out directly myself and will have to try to troubleshoot this remotely with your help.

re: outtest1, if you stop the code after the `gettagpr()` call then I would definitely expect the output to be different, as the file won't have been completely written at that point.  What I'd be really curious to see is a copy of the full out1.bufr output file that results from completely running through that test all the way up through the `stop 5` statement, b/c that way I could dig into it to see where/how it really differs from what we'd expect.  Is there any way you could send that to me as a file attachment?  It should be in the test subdirectory of your local build_gcc14 directory.

Sorry but off the top of my head nothing jumps out as to what those particular failing test codes may have in common, so again I think I'm going to need to brute force my way through this, starting with out1.bufr.  If you can also send me the corresponding out11.bufr, out12.bufr, and out13.bufr files then I'll try to look at those as well.

**Comment by AlexanderRichert-NOAA** [comment]
Yes, I can send the resulting files for the four failing tests. Also, I should have clarified: when I said the output was different at the early point in the test, I meant different *between the GCC 12 and GCC 14 builds*. So whatever is going on doesn't seem to be down to any one subroutine in the series that is getting called.

**Comment by AlexanderRichert-NOAA** [comment]
Here's a zip file with out{1,11,12,13}.bufr based on GCC 14, running the _d versions of those tests: [failures.zip](https://github.com/user-attachments/files/19395161/failures.zip) (sha256 sum: 843f9b4e96c1ad0f25a9328e7a33645fb529f3e186780039f2af9b58ed6d808d)

**Comment by jbathegit** [comment]
Thanks for the zipfile Alex!  I spent some time digging through those files, and they're definitely all wrongly encoded, and which in turn is why those particular tests are now failing (as they should be :-)

Unfortunately I'm not seeing any real patterns that give me any clue as to what the problem(s) may be.  I do see that most of the tests which are failing seem to be writing compressed output, and a few of them also seem to be writing out "long" (i.e. longer than 8 byte) character strings, but that still doesn't give me enough to go on to really nail this down.  What I'd like to do is step through those particular test codes (and the parts of the library they call) with a debugger, to see where specifically the output is diverging from what we see in earlier gcc12 and gcc13 builds.  But to do that I may just have to wait until a gcc14 compiler becomes available locally on WCOSS2, hopefully in the not-too-distant future.

So my suggestion for now is to leave this issue open until then, and unless you (or @jack-woollen, or anyone else) wants to try to tackle it yourself in the meantime on some other platform which already has a gcc14 compiler(?)
